### ISSUE TITLE
policies/groot: the "lerobot ships it upstream" rationale holds for local mode only - service mode and the container tool have no upstream replacement, so removing the provider is a decision, not a cleanup

### ISSUE THREAD

**Issue by cagataycali** [strands-labs/robots#3075]
#3047 proposed removing `strands_robots/policies/groot/` whole, on the stated rationale that "lerobot ships both upstream, so strands-robots carrying its own copies is duplicated surface a reader has to disambiguate." That rationale was correct for the LIBERO half, which landed on its own in #3056.

I measured the same claim against the GR00T half before rescoping #3047 onto it, and it only holds for part of the provider. Filing the finding here rather than acting on it in a PR, because what is left is a decision about a documented public provider and not a cleanup.

## The premise is true

LeRobot really does ship a GR00T policy upstream - `src/lerobot/policies/groot` is present in `huggingface/lerobot` at HEAD, and this repo already depends on that fact in two places:

- `strands_robots/policies/lerobot_local/resolution.py:364` avoids `lerobot.policies.__init__` because "in LeRobot 0.5+ eagerly imports groot -> transformers".
- `strands_robots/policies/lerobot_async/policy.py:77` carries `"groot"` in `_SUPPORTED_POLICY_TYPES_FALLBACK`, a snapshot of lerobot's own `SUPPORTED_POLICIES`.

So "lerobot ships groot" is not in doubt. What is in doubt is that it ships *this* provider's surface.

## What the provider actually exposes, and what upstream replaces

`Gr00tPolicy` selects its mode on `model_path` (`policy.py:602`): set it and the model is loaded in-process, leave it `None` and the policy dials a server.

| Surface | What it does | Upstream replacement |
|---|---|---|
| local mode (`model_path=...`) | loads GR00T weights in-process | **yes** - lerobot's `policies/groot` via `lerobot_local` |
| service mode (`host`/`port`, default) | dials **Isaac-GR00T's own ZMQ inference server** | **no** - `lerobot_async` speaks lerobot's gRPC `PolicyServer`, a different server and a different wire protocol |
| `tools/gr00t_inference.py` | Isaac-GR00T Docker lifecycle: `build_image` / `download_checkpoint` / `start_container` | **no** |
| `groot/data_configs.json` | 27 embodiment configs + 3 aliases, keyed to Isaac-GR00T's modality-config shape | **no** - these describe the Isaac-GR00T server's payload, not a lerobot policy's |

Service mode is the default: `create_policy("groot", port=5555)` is the form the docs lead with, and it is the form that needs no GPU in the agent process. Substituting `lerobot_local` there does not preserve the deployment - it moves inference into the caller. Substituting `lerobot_async` changes which server the operator has to be running.

So removing the provider whole drops a capability rather than deduplicating one. That is the part #3047's rationale does not cover, and it is why I did not rescope #3047 onto it.

## What a removal would break

- `create_policy("groot", ...)` and the `"groot"` spelling in `registry/policies.json` (`url_patterns: ["^zmq://"]`, `hf_orgs: ["nvidia"]`, `model_id_overrides`)
- the `[groot-service]` extra
- `Gr00tTrainer` (`training/groot.py`) and the `gr00t_inference` AgentTool
- `docs/policies/groot.md`, plus `groot` references in about 20 further doc files
- `mesh.security._REGISTRY_POLICY_PROVIDERS`, which is documented as every spelling in `registry/policies.json`

## Options

1. **Keep the provider.** The duplicated part is local mode only, and it is a few branches inside a module whose main job is the ZMQ client. Optionally document that `lerobot_local` with `policy_type="groot"` is the in-process route, so the two are disambiguated in prose rather than by deletion - which is the problem the rationale actually names.
2. **Narrow the removal to local mode**, keeping service mode, the data configs and the container tool. This deduplicates exactly what upstream covers and keeps the deployment shape. Needs a deprecation window for `model_path=`.
3. **Remove the provider whole**, accepting the loss of service-mode Isaac-GR00T and the container tool, with a migration note pointing at `lerobot_async`. This is a breaking change to a documented provider and wants a deprecation release first, not a straight delete.

I have no stake in which; option 1 is the status quo and costs nothing, and options 2 and 3 both need a maintainer to accept a public-API change. #3047 is closed rather than rescoped so this is not decided by a draft nobody asked for - the branch is kept for its measured work (the 11 `policy_provider=` retargets, the lock regeneration, the shared-table retargets) if option 3 is ever chosen.

The grading concern raised on #3047's `tool_spec.json` thread is separately resolved on `main` by #3054, which made the unbacked-key suite grade both answers a provider can give off the registry instead of assuming absorption. No `Cosmos3Policy` change is needed for it, and none was made.

**Comment by cagataycali** [comment]
Agreed on the framing, and I reached #3054 independently while measuring the same thing -- the unbacked-key suite grades both answers off the registry now, so nothing is owed on `Cosmos3Policy`.

Two corrections and some cost data for option 3, from carrying the whole removal far enough to see what it touches.

**The `policy_provider=` retarget count is 17, not 11**, and there is an 18th site of a different kind:

| site | count |
|---|---|
| `hardware_robot.Robot` -- `start_task` plus 5 private task helpers | 6 |
| native drivers' `start_task` -- `base`, `booster`, `crazyflie`, `franka`, `g1`, `go2`, `microduck`, `reachy`, `robotiq`, `ur` | 10 |
| `tools/g1/g1_start_task.py` | 1 |
| `"default": "groot"` inside `Robot.tool_spec`'s `policy_provider` property | 1 (schema, not signature) |

The schema one matters separately because it is written by hand next to the signature default rather than derived from it, so the two go stale independently -- an agent reads the schema, a Python caller reads the signature.

**Those defaults constrain option 3 more than the count suggests.** `start_task` documents `policy_port` as required precisely because "this entry point takes no pre-built policy, so the port is the only thing a policy can be built from", so the defaulted provider *is* what gets built, and it is reached by exactly the calls that pass the fewest arguments. That rules out the obvious substitution: `lerobot_local` builds in-process and would silently ignore `policy_port`, turning a port-dialing default into one that discards the caller's only means of specifying a policy. The replacement has to declare a port, which among survivors means `cosmos3` / `lerobot_async` / `remote` / `vera` -- and the registry's `requires` field is already documented as the source of truth for that distinction, in `_get_policy`'s own comment ("groot/lerobot_async dial a server and need a port, while mock/lerobot_local build in-process and need none").

**Cost of option 3, measured rather than estimated.** With the provider unregistered, the extra dropped and every executable reference repointed, the whole-tree graders do go green (177 passed, 1 skipped), and getting there is the bulk of the work because a removal empties the *cited* side of every cross-reference while a grader's population is the tree rather than the diff:

| grader | what it turned up |
|---|---|
| `test_docstring_xref_roles_resolve` | 10 dead `:mod:`/`:class:` roles, in `policies/cosmos3`, `policies/curobo`, `policies/moveit2` (3 sites), `training/sagemaker`, `utils`, and 2 test modules |
| `test_dependency_audit` (both cells) | 7 install hints and table rows naming `[groot-service]`, across `README.md`, `docs/architecture.md`, `docs/getting-started/installation.md`, `docs/policies/overview.md`, `docs/training/vla_workflow.md`, an example and a test |
| `test_agent_tool_parameter_descriptions` | its pinned tool inventory, plus a whole class whose subject is the removed tool's container-lifecycle options |

The residual is the part that makes option 3 more than a delete: **12 test modules import `Gr00tTrainer` or `Gr00tPolicy` directly (8 collection errors)**, and 7 of those are *shared-domain* guards that enumerate the trainer in a parametrize tuple -- `test_seed_domain`, `test_run_size_domain`, `test_learning_rate_domain`, `test_launch_topology_domain`, `test_checkpoint_cadence_domain`, `test_validation_episodes_domain`, `test_lora_hyperparameter_domain`. `test_zmq_timeout_ms_domain` is the sharpest: it grades the shared `coerce_zmq_timeout_ms` domain across *both* ZMQ REQ inference clients, and only `MoveIt2InferenceClient` would survive, so that docstring's "both clients" reasoning has to be restated rather than trimmed. Those want re-anchoring on the merits, the way #3056 re-anchored its three.

Two further consequences of option 3 worth pricing in, since neither is a reference update:

- `examples/locomotion/vla_g1_workflow.py` calls `create_trainer("groot")` as its stage 2, and `docs/training/vla_workflow.md` documents that stage. Stages 1 and 3 survive intact -- notably the WBC provider still consumes a GR00T-WBC (SONIC) checkpoint, so `create_policy("wbc", checkpoint=".../grootwbc-g1")` is unaffected -- but the page's three-stage thesis becomes two, with post-training moved to upstream tooling.
- `docs/policies/groot.md` is in the `mkdocs.yml` nav and is linked from 8 other pages, so the nav entry and those links go with it.

**Independent of which option is chosen**, the gap you name -- "nothing links a registry entry to the defaults that name it" -- is worth closing on its own, because it is what makes the 18 sites above invisible until a caller omits the argument. A derived guard covers it cheaply: walk the package for `policy_provider` defaults (signature, via AST) and for `policy_provider` `"default"` literals in tool schemas, and assert each names a registered provider, skipping `None` as the not-supplied sentinel. Deriving the population rather than listing it means a provider-dialing method added later is held to the rule on arrival. It passes on `main` today, which is the point -- it is a floor for any future provider rename or removal, and under option 1 it costs nothing and still pays.

Happy to bring that guard up on its own if it is wanted; it does not presuppose any of the three options.

**Comment by cagataycali** [comment]
The decision-independent guard offered at the end of this issue is up as #3183, and it turned out to have a live instance rather than being the pure floor I predicted. Recording that, plus two corrections to the sizing above.

## The guard was not passing on `main`

My own words were "it passes on `main` today, which is the point - it is a floor for any future provider rename or removal". Measured on `2f37927`, the *signature-default* half does pass, all 30 sites. The half I described as the same walk's cheap other side does not:

| site | advertises | registered? |
|---|---|---|
| `simulation/mujoco/tool_spec.json` | `groot, lerobot_local, lerobot_async, cosmos3, mock` | 5 of 5 |
| `hardware_robot.py:2655` (`Robot.tool_spec`) | `groot, openai` | **1 of 2** |

`openai` is in none of the 31 spellings the two registries hold. `resolve_policy('openai')` returns `('lerobot_local', {'pretrained_name_or_path': 'openai'})`, and on the hardware path `create_policy` raises `Unknown policy provider` from `_get_policy` on the executor thread - after `start_task` has already answered `Task started` and the arm has connected.

The sharper symptom is the one this issue's `requires` paragraph predicts without naming it. `_get_policy` falls back to demanding a port when the provider is unknown, so the schema's own name passed without a port is refused for the *port*:

```
start_task(..., policy_provider='openai') -> start_task: policy_port is required to build a policy
start_task(..., policy_provider='groot')  -> start_task: policy_port is required to build a policy
```

Byte-identical. So the conservative fallback that keeps port-less providers runnable also makes an unregistered provider indistinguishable from a registered port-dialing one, and nothing on that path says the name does not exist.

## Correction to the retarget count: 30 signature defaults, not 17

The table above counts 17 `policy_provider=` sites plus the schema literal. Measured by AST over the package:

| default | count |
|---|---|
| `"groot"` | 18 |
| `"mock"` | 10 |
| `None` (not-supplied sentinel) | 2 |

The `groot` group is 18 rather than 17 because the table's native-driver list omits `drivers/earthrover.py:470`. The 10 `mock` sites are the simulation and device-connect rollout helpers (`simulation/base.py` x4, `simulation/mujoco/simulation.py` x2, `device_connect/{sim,robot}_driver.py`, `tools/run_policy.py`, `tools/robot_mesh.py`); they are not in scope for a `groot` removal, but they are in scope for the guard, which is why the sweep grades the population rather than the `groot` subset.

The schema-literal count of 1 is confirmed - `hardware_robot.py:2658` is the only `"default"` under a `policy_provider` property in the tree. The MuJoCo tool spec carries `description` and `type` only, no `default`.

## What #3183 does and does not touch

It implements the guard as described here, with one change of shape: rather than a new module it widens `tests/registry/test_advertised_providers_are_registered.py`, which already exists for this exact class - its docstring names both the late `Unknown policy provider` and the silent `lerobot_local` redirect - and whose population was one file, the MuJoCo tool spec. That is why the second surface was reachable with the suite green, and it is the same shape as the census in #2791: a grader that names its own subject cannot see a sibling arrive. The population is now derived, so a third schema surface is graded on arrival.

Options 1-3 are untouched. I have unregistered nothing, removed no extra and repointed no default, so the cost table above stands as the pricing for whichever is chosen - and the 30 sites are now graded, which is what makes a repoint a visible edit rather than one that goes quiet until a caller omits the argument.

Leaving this issue open for the removal decision, which is the part that still wants a maintainer.

**Comment by cagataycali** [comment]
Decision recorded: GR00T stays. Service mode and the container tool have no upstream replacement (this issue's own measurement), and the provider cleanup direction keeps GR00T/MolmoAct while dropping lerobot_async, curobo, moveit2 and motionbricks. Nothing further to do here.
